### PR TITLE
[TC-187] fix HTTPs bugs

### PR DESCRIPTION
**Pull request by zhilhuan**
use ds id as key to store ssl certificate in riak server;
update ssl certificate in keystore if delivery service modified xml-id or subdomain;
repose decoded certificate in restful api.

### CODE REVIEW

**Comment by rawlinp** [comment]
@zhilhuan can you please rebase this PR?

This PR also seems to fix an issue we'll have once per-DeliveryService routing names are implemented soon. Basically, if a DNS DS starts using the DNS routing name "my-edge" rather than the default "edge", the new hostname will mismatch the hostname used in the SSL cert. Your PR seems to address that issue, which is helpful in addition to the issue of changing the xml_id.

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by zhilhuan** [comment]
@rawlinp rebased.

Yes, we found several SSL cert issues related to update delivery service xml_id or sub-domain. And fixed those issue in this PR.

File: traffic_ops/app/lib/UI/DeliveryService.pm [hunk 1]
```diff
@@ -846,8 +842,11 @@ sub update {
 			$hash{http_bypass_fqdn} = $self->param('ds.http_bypass_fqdn');
 		}
 
+		my $upd_ssl = 0;
 		#print Dumper( \%hash );
 		my $update = $self->db->resultset('Deliveryservice')->find( { id => $id } );
+		$upd_ssl = 1 if $update->xml_id ne $hash{xml_id};
```
Comment (DylanVolz) [inline traffic_ops/app/lib/UI/DeliveryService.pm]: This line is may be  unnecessary now that [TC-402] was merged.

Comment (zhilhuan) [inline traffic_ops/app/lib/UI/DeliveryService.pm]: OK, will remove this line

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/26/

File: traffic_ops/app/lib/API/DeliveryService/SslKeys.pm [hunk 2]
```diff
@@ -161,30 +172,42 @@ sub view_by_hostname {
...
 	if ( !&is_admin($self) ) {
 		return $self->alert( { Error => " - You must be an ADMIN to perform this operation!" } );
 	}
 	else {
+		my $ds = $self->db->resultset('Deliveryservice')->search( { xml_id => $xml_id })->single();
+		if (!$ds) {
+			return $self->alert( { Error => " - Could not found delivery service with xml_id=$xml_id!" } );
+		}
+		my $ds_id = $ds->id;
+		my $key = "ds_$ds_id";
```
Comment (rawlinp) [inline traffic_ops/app/lib/API/DeliveryService/SslKeys.pm]: preexisting keys won't be deletable?

File: traffic_ops/app/lib/UI/DeliveryService.pm [hunk 3]
```diff
@@ -152,12 +153,7 @@ sub get_example_urls {
 				}
 			}
 			elsif ( $re->{type} eq 'PATH_REGEXP' ) {
-				if ( defined( $example_urls[ $re->{set_number} ] ) ) {
-					$example_urls[ $re->{set_number} ] .= $re->{pattern};
-				}
-				else {
-					$example_urls[ $re->{set_number} ] = $re->{pattern};
-				}
+				push(@example_urls, $re->{pattern});
```
Comment (rawlinp) [inline traffic_ops/app/lib/UI/DeliveryService.pm]: What made this change necessary? Doesn't the old code concatenate patterns if they use the same `set_number`? This looks like it would just add multiple different patterns using the same `set_number`s.

File: traffic_ops/app/script/update_riak_for_search.pl [hunk 4]
```diff
@@ -46,9 +46,6 @@
 		}
 		$record->{deliveryservice} = $xml_id;
 		$record->{cdn} = $cdn;
-		$record->{certificate}->{crt} = decode_base64($record->{certificate}->{crt});
```
Comment (rawlinp) [inline traffic_ops/app/script/update_riak_for_search.pl]: This is an example of a client that would break, if this particular API is being used by a client outside of our codebase.

File: traffic_ops/app/lib/UI/SslKeys.pm [hunk 5]
```diff
@@ -75,6 +75,34 @@ sub add {
 	}
 }
 
+sub update_sslkey {
+	my $self = shift;
+	my $ds_id = shift;
+	my $xml_id = shift;
+	my $hostname = shift;
+	my $response_container = $self->riak_get( 'ssl', "ds_$ds_id-latest");
```
Comment (rawlinp) [inline traffic_ops/app/lib/UI/SslKeys.pm]: preexisting keys can't be updated?

File: traffic_ops/app/lib/UI/SslKeys.pm [hunk 6]
```diff
@@ -34,7 +34,7 @@ sub add {
 	&stash_role($self);
 
 	#get key data from keystore
-	my $response_container = $self->riak_get( 'ssl', "$xml_id-latest");
+	my $response_container = $self->riak_get( 'ssl', "ds_$ds_id-latest");
```
Comment (rawlinp) [inline traffic_ops/app/lib/UI/SslKeys.pm]: a preexisting key won't be found and won't get stashed. I'm not sure how much of an issue that might be.

File: traffic_ops/app/lib/UI/DeliveryService.pm [hunk 7]
```diff
@@ -921,6 +919,10 @@ sub update {
 			}
 		}
 
+		my $new_hostname = UI::SslKeys::get_hostname($self, $id, $update);
+		$upd_ssl = 1 if $old_hostname ne $new_hostname;
```
Comment (rawlinp) [inline traffic_ops/app/lib/UI/DeliveryService.pm]: This is good, but shouldn't the API also perform the same check and update SSL keys if the hostname changed?

File: traffic_ops/app/lib/API/DeliveryService/SslKeys.pm [hunk 8]
```diff
@@ -120,12 +120,23 @@ sub view_by_xml_id {
 		if ( !$version ) {
 			$version = 'latest';
 		}
-		$key = "$key-$version";
+		my $ds = $self->db->resultset('Deliveryservice')->search( { xml_id => $xml_id })->single();
+		if (!$ds) {
+			return $self->alert( { Error => " - Could not found delivery service with xml_id=$xml_id!" } );
+		}
+		my $ds_id = $ds->id;
+		my $key = "ds_$ds_id-$version";
```
Comment (zhilhuan) [inline traffic_ops/app/lib/API/DeliveryService/SslKeys.pm]: As discussed in Slack, xml_id is immutable now, will revert this change in TC-547.

File: traffic_ops/app/lib/API/DeliveryService/SslKeys.pm [hunk 9]
```diff
@@ -120,12 +120,23 @@ sub view_by_xml_id {
...
 		my $response_container = $self->riak_get( "ssl", $key );
 		my $response = $response_container->{"response"};
-		$response->is_success()
-			? $self->success( decode_json( $response->content ) )
-			: $self->alert( { Error => " - A record for ssl key $key could not be found.  Response was: " . $response->content } );
+		if ($response->is_success()) {
+			my $ssl_keys = decode_json( $response->content );
+			$ssl_keys->{certificate}->{csr} = decode_base64($ssl_keys->{certificate}->{csr}),
```
Comment (zhilhuan) [inline traffic_ops/app/lib/API/DeliveryService/SslKeys.pm]: We made this change because it is not compatible with GUI interface. Since this may case regress issue, will revert it back in TC-547.

File: traffic_ops/app/lib/API/DeliveryService/SslKeys.pm [hunk 10]
```diff
@@ -161,30 +172,42 @@ sub view_by_hostname {
...
 		}
 
-		my $xml_id = $ds->xml_id;
+		my $ds_id = $ds->id;
 
 		if ( !$version ) {
 			$version = 'latest';
 		}
-		$key = "$xml_id-$version";
+		$key = "ds_$ds_id-$version";
```
Comment (zhilhuan) [inline traffic_ops/app/lib/API/DeliveryService/SslKeys.pm]: As discussed in Slack, xml_id is immutable now, will revert this change in TC-547.

File: traffic_ops/app/lib/API/DeliveryService/SslKeys.pm [hunk 11]
```diff
@@ -161,30 +172,42 @@ sub view_by_hostname {
...
 		my $response_container = $self->riak_get( "ssl", $key );
 		my $response = $response_container->{"response"};
-		$response->is_success()
-			? $self->success( decode_json( $response->content ) )
-			: $self->alert( { Error => " - A record for ssl key $key could not be found.  Response was: " . $response->content } );
+		if ($response->is_success()) {
+			my $ssl_keys = decode_json( $response->content );
+			$ssl_keys->{certificate}->{csr} = decode_base64($ssl_keys->{certificate}->{csr}),
```
Comment (zhilhuan) [inline traffic_ops/app/lib/API/DeliveryService/SslKeys.pm]: will revert this change in TC-547.

File: traffic_ops/app/lib/API/DeliveryService/SslKeys.pm (hunk 2 above)
Comment (zhilhuan) [inline traffic_ops/app/lib/API/DeliveryService/SslKeys.pm]: will revert this change in TC-547.

File: traffic_ops/app/lib/UI/DeliveryService.pm (hunk 3 above)
Comment (zhilhuan) [inline traffic_ops/app/lib/UI/DeliveryService.pm]: It messed up when both "HOST_REGEXP" and "PATH_REGEXP" configured. Root cause:

1. Duplicated "set_number" is not allowed in DS configuration in GUI. So not "PATH_REGEXP" will share the same "set_number" with "HOST_REGEXP".
2. "$re->{set_number}" is not the index of "@example_urls". See the "push( @example_urls, $https_url );" code above. So "$example_urls[ $re->{set_number} ] .= $re->{pattern};" may mess up with any other "HOST_REGEXP".

File: traffic_ops/app/lib/UI/DeliveryService.pm (hunk 7 above)
Comment (zhilhuan) [inline traffic_ops/app/lib/UI/DeliveryService.pm]: good catch! I think the code change is missing when doing merge.

The original change is made in TC 1.7. When picked up to master, looks like the code is lost :-(

Comment (zhilhuan) [inline traffic_ops/app/lib/UI/DeliveryService.pm]: Will also add it to TC-547

File: traffic_ops/app/lib/UI/SslKeys.pm (hunk 6 above)
Comment (zhilhuan) [inline traffic_ops/app/lib/UI/SslKeys.pm]: will revert this change in TC-547.

File: traffic_ops/app/lib/UI/SslKeys.pm (hunk 5 above)
Comment (zhilhuan) [inline traffic_ops/app/lib/UI/SslKeys.pm]: will use xml_id instead in TC-547.

File: traffic_ops/app/script/update_riak_for_search.pl (hunk 4 above)
Comment (zhilhuan) [inline traffic_ops/app/script/update_riak_for_search.pl]: will revert this change in TC-547.
